msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AadInstanceDiscoveryProvider.java

```diff
@@ -62,40 +62,39 @@ static InstanceDiscoveryMetadataEntry getMetadataEntry(URL authorityUrl,
                                                            boolean validateAuthority,
                                                            MsalRequest msalRequest,
                                                            ServiceBundle serviceBundle) {
-        String host = authorityUrl.getHost();
-
-        if (shouldUseRegionalEndpoint(msalRequest)) {
-            //Server side telemetry requires the result from region discovery when any part of the region API is used
-            String detectedRegion = discoverRegion(msalRequest, serviceBundle);
 
-            if (msalRequest.application().azureRegion() != null) {
-                host = getRegionalizedHost(authorityUrl.getHost(), msalRequest.application().azureRegion());
-            }
+        String host = authorityUrl.getHost();
 
-            //If region autodetection is enabled and a specific region not already set,
-            // set the application's region to the discovered region so that future requests can skip the IMDS endpoint call
-            if (null == msalRequest.application().azureRegion() && msalRequest.application().autoDetectRegion()
-                    && null != detectedRegion) {
-                msalRequest.application().azureRegion = detectedRegion;
-            }
-            cacheRegionInstanceMetadata(authorityUrl.getHost(), msalRequest.application().azureRegion());
-            serviceBundle.getServerSideTelemetry().getCurrentRequest().regionOutcome(
-                    determineRegionOutcome(detectedRegion, msalRequest.application().azureRegion(), msalRequest.application().autoDetectRegion()));
+        //If instanceDiscovery flag set to false or instance discovery previously failed, do not do instance discovery
+        if (!msalRequest.application().instanceDiscovery() || instanceDiscoveryFailed) {
```

Comment:
nit: it might be simpler to do smth like: 

- if a cached entry exists, use it
- if not and instance discovery has been disabled, add the 1-host dumb entry to the cache. Same if instance discovery fails.

makes the code shorter and removes the need for having `instanceDiscoveryFailed` flag

Reply:
In the latest commit I rearranged it so this is now the pattern:
1. If instance discovery is disabled, just add a default entry (if it doesn't already exist) and return it
2. If a region is set at the application level, adjust the `host` variable to use it
3. If no cached entry for `host` exists, then do instance discovery
4. If any part of the region API is used, try to detect the region
5. If region autodetect enabled and a region wasn't set at the application level, set the region at the application level and adjust the `host` to use it
6. Do instance discovery for `host`, and add the default entry if it fails
7. Create cache entries for whatever `host` is at the end of all that
8.  Return the cached result

This means on future runs the cached entry will be returned quicker, because:
- Instance discovery was disabled, so step 1 returns immediately
- Instance discovery failed and a default entry for `host` was cached, so it skips steps 3-7 and returns the cached entry
- Instance discovery succeeded and either a regional or non-regional entry for `host` was cached, so it skips steps 3-7 and returns the cached entry